### PR TITLE
Support for VectorCoefficient and MatrixCoefficient with VectorDiffusionIntegrator [wcdawn/nonscalar-VectorDiffusionCoeficient]

### LINKED ISSUE #1637: System of Diffusion-like Equations (Multigroup Neutron Diffusion)
Background
---

I'm working to solve the multigroup neutron diffusion equations.
It is essentially a coupled system of diffusion equations that represents an eigenvalue problem for eigenvalue \lambda and eigenvector \phi_g(\mathbf{r}).

\nabla \cdot \left( D_g(\mathbf{r}) \nabla \phi_g(\mathbf{r}) \right) + \Sigma_{t,g}(\mathbf{r}) \phi_g(\mathbf{r}) = \frac{\chi_g(\mathbf{r})}{\lambda}  \sum_{g'=1}^{G} \nu \Sigma_{f,g'}(\mathbf{r}) \phi_{g'}(\mathbf{r}) + \sum_{g'=1}^{G} \Sigma_{s, g' \rightarrow g}(\mathbf{r}) \phi_{g'}(\mathbf{r}) \quad \text{for } g = 1,2,3, \ldots, G

This system can be rewritten as a generalized eigenvalue problem.

\mathbf{A} \Phi = \frac{1}{\lambda} \mathbf{M}

I am only interested in the eigenmode with largest eigenvalue \lambda. It has been shown elsewhere that the most efficient and robust solution method for this class of equations is the Generalized Davidson eigensolver and I'm currently using the implementation in SLEPc (for a one-group, G=1, problem). A Power Iteration algorithm has been simple to implement for the multigroup problem (G\ge1) but the method itself is rather inefficient.

Problem
---

The \mathbf{A} and \mathbf{M} matrices from above have dimensions \mathbf{A}, \mathbf{M} \in \mathbb{R}^{NG \times NG} where N is the DOF of the mesh and G is the total number of energy groups. Essentially, each grid position has G functions associated with it.

To minimize bandwidth, it is favorable to arrange the matrix so that the matrix index is i=g + Gn where n is the index for the original DOF (as possible by `Ordering::byVDIM`). This seemed like a great candidate for the `Vector*` classes (e.g. `VectorCoefficient`, `VectorMassIntegrator`, etc.) in MFEM as there is a vector of functions associated with each grid location. However, this did not seem to work as I imagined, particularly the `VectorDiffusionIntegrator`.

It seems it may be possible to write my own `BilinearIntegrator` but I would like to leverage the existing code if possible. Especially as this will be scaling to very large supercomputers.

How should I go about building these \mathbf{A} and \mathbf{M} matrices in MFEM?

Thank you in advance!

### PR DESCRIPTION
**Pull request by wcdawn**
Add support for nonscalar coefficients (`VectorCoefficient` and `MatrixCoefficient`) for use with class `VectorDiffusionCoefficient`.

This was needed to model multigroup neutron diffusion related to #1637.

Resolves #1637 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1727](https://github.com/mfem/mfem/pull/1727) | @wcdawn | @tzanio | @stefanozampini + @rw-anderson | 08/30/20 | 01/14/21 | 01/27/21 | |

### CODE REVIEW

File: fem/bilininteg.cpp [hunk 1]
```diff
@@ -1,3 +1,7 @@
+// vim: ts=3 sw=3 sts=3
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:1]: This code block does not belong to the PR

File: fem/bilininteg.cpp [hunk 2]
```diff
@@ -2174,20 +2178,31 @@ void VectorDiffusionIntegrator::AssembleElementMatrix(
    const int dim = el.GetDim();
    const int dof = el.GetDof();
    const int sdim = Trans.GetSpaceDim();
-   const bool square = (dim == sdim);
-   double w;
 
-   elmat.SetSize(sdim * dof);
+   // If vdim is not set, set it to the space dimension;
+   vdim = (vdim <= 0) ? sdim : vdim;
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2183]: What if `VQ` or `MQ` have a different` vdim`?  You should probably add some description in the hpp file

File: fem/bilininteg.cpp [hunk 3]
```diff
@@ -2247,29 +2281,43 @@ void VectorDiffusionIntegrator::AssembleElementVector(
...
    for (int i = 0; i < ir->GetNPoints(); i++)
    {
       const IntegrationPoint &ip = ir->IntPoint(i);
+      el.CalcDShape(ip, dshape);
 
       Tr.SetIntPoint(&ip);
       CalcAdjugate(Tr.Jacobian(), Jinv);
-      w = ip.weight / Tr.Weight();
-      if (Q)
+      double w{ip.weight / Tr.Weight()};
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2301]: use `=` for assignments? applies to all the curly brackets of this code

File: fem/bilininteg.cpp [hunk 4]
```diff
@@ -2247,29 +2281,43 @@ void VectorDiffusionIntegrator::AssembleElementVector(
...
+
+      if (VQ)
       {
-         w *= Q->Eval(Tr, ip);
+         mfem_error("VectorDiffusionIntegrator::AssembleElementVector \n"
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2305]: Why raising en error? can't you provide the application as done for the matrix assembly?

Comment (wcdawn) [reply fem/bilininteg.cpp:2305]: I don't understand what this routine is supposed to do, I don't need it for my work, and there are no examples in any of examples, miniapps, or unittests. I'll try to implement it but I have no way of testing.

Comment (stefanozampini) [reply fem/bilininteg.cpp:2305]: `AssembleElementVector` is the routine that computes the FEM residual. For linear problems, 
```
elvect = elmat * elfun
```
 where `elfun` is the input linearization point .

Comment (wcdawn) [reply fem/bilininteg.cpp]: I've implemented with a different style than the original function but in the same style as `AssembleElementMatrix`.

**Comment by tzanio** [comment]
@stefanozampini and @rw-anderson -- your reviews on this PR are overdue

**Comment by tzanio** [comment]
Second reminder: @stefanozampini and @rw-anderson -- your reviews on this PR are overdue

File: fem/bilininteg.cpp [hunk 5]
```diff
@@ -2174,20 +2174,31 @@ void VectorDiffusionIntegrator::AssembleElementMatrix(
...
+
+   // If vdim is not set, set it to the space dimension;
+   vdim = (vdim <= 0) ? sdim : vdim;
    const bool square = (dim == sdim);
-   double w;
 
-   elmat.SetSize(sdim * dof);
+   if (VQ)
+   {
+      vcoeff.SetSize(vdim);
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2184]: I see this the other way around. It should be the vector or matrix coefficient to set `vdim` on line 2179. The integrators should not be in charge of modifying the coefficients.

File: fem/bilininteg.cpp [hunk 6]
```diff
@@ -2199,55 +2210,91 @@ void VectorDiffusionIntegrator::AssembleElementMatrix(
...
+   const int dof = el.GetDof();
+   const int sdim = Tr.GetSpaceDim();
+
+   // If vdim is not set, set it to the space dimension;
+   vdim = (vdim <= 0) ? sdim : vdim;
+   const bool square = (dim == sdim);
+
+   if (VQ)
+   {
+      vcoeff.SetSize(vdim);
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2276]: Same here, see comment above

File: fem/bilininteg.cpp [hunk 7]
```diff
@@ -2199,55 +2210,91 @@ void VectorDiffusionIntegrator::AssembleElementMatrix(
...
+      mcoeff.SetSize(vdim);
+   }
 
-   Jinv.SetSize(dim);
    dshape.SetSize(dof, dim);
    pelmat.SetSize(dim);
-   gshape.SetSize(dim);
 
    elvect.SetSize(dim*dof);
+   // NOTE: DenseMatrix is in column-major order.
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2287]: I guess you don't need this comment here, but it is a matter of taste

File: fem/bilininteg.cpp [hunk 8]
```diff
@@ -2257,26 +2304,51 @@ void VectorDiffusionIntegrator::AssembleElementVector(
...
+      double w = Tr.Weight();
+      w = ip.weight / (square ? w : w*w*w);
+
+      if (VQ)
       {
-         w *= Q->Eval(Tr, ip);
+         VQ->Eval(vcoeff, Tr, ip);
+         for (int k = 0; k < vdim; ++k)
+         {
+            Mult_a_AAt(w*vcoeff(k), dshapedxt, pelmat);
```
Comment (stefanozampini) [inline fem/bilininteg.cpp:2318]: I guess you don't need to construct `pelmat`, only to provide it's action 
```
vec_out += w*coeff(k)*dshapeext * dshapeext^T vec_in
```

File: fem/bilininteg.hpp [hunk 9]
```diff
@@ -2519,14 +2519,22 @@ class DivDivIntegrator: public BilinearFormIntegrator
...
-    for FE spaces defined by 'dim' copies of a scalar FE space. Where e_i
-    is the unit vector in the i-th direction.  The resulting local element
-    matrix is a block-diagonal matrix consisting of 'dim' copies of a scalar
-    diffusion matrix in each diagonal block. */
+    for vector FE spaces. Where e_i is the unit vector in the i-th direction.
+    The resulting local element matrix is square with dimension
+    <tt> vdim * dof </tt> where \c vdim is the vector dimension space and \c dof
+    is the local degrees of freedom. The integrator is not aware of the true
+    vector dimension and must use the \c VectorCoefficient,
+    \c MatrixCoefficient, or a caller-specified value to determine the vector
```
Comment (stefanozampini) [inline fem/bilininteg.hpp:2527]: Here you say the integrator may inherit the block size from the coefficients, but in practice it does not

File: fem/bilininteg.hpp [hunk 10]
```diff
@@ -2536,11 +2544,50 @@ class VectorDiffusionIntegrator : public BilinearFormIntegrator
...
+
+   /** \brief Integrator with scalar coefficient for caller-specified vector dimension.
+
+      The element matrix is block-diagonal with \c vdim copies of the element
+      matrix integrated with the \c Coefficient.
+
+      If the vector demension does not match the true dimension of the \c
+      FiniteElementSpace , the resulting element matrix will be mathematically
+      invalid. */
+   VectorDiffusionIntegrator(Coefficient &q, int vector_dimension) : Q{&q}, vdim{vector_dimension} {}
```
Comment (stefanozampini) [inline fem/bilininteg.hpp:2571]: Matter of taste, but in MFEM I guess we prefer using parenthesis for the constructors

```
: Q(&q), vdim(vector_dimension) {}
```

File: fem/bilininteg.hpp [hunk 11]
```diff
@@ -2536,11 +2544,50 @@ class VectorDiffusionIntegrator : public BilinearFormIntegrator
...
+
+   /** \brief Integrator with \c VectorCoefficient. The vector dimension of the \c FiniteElementSpace is assumed to be the same as the dimension of the \c Vector.
+
+      The element matrix is block-diagonal and each block is integrated with coefficient q_i.
+
+      If the vector demension does not match the true dimension of the \c
+      FiniteElementSpace , the resulting element matrix will be mathematically
+      invalid. */
+   VectorDiffusionIntegrator(VectorCoefficient& vq) : VQ{&vq}, vdim{vq.GetVDim()} {}
```
Comment (stefanozampini) [inline fem/bilininteg.hpp:2580]: same as above

File: fem/bilininteg.hpp [hunk 12]
```diff
@@ -2536,11 +2544,50 @@ class VectorDiffusionIntegrator : public BilinearFormIntegrator
...
+
+   /** \brief Integrator with \c MatrixCoefficient. The vector dimension of the \c FiniteElementSpace is assumed to be the same as the dimension of the \c Vector.
+
+      The element matrix is populated in each block.
+      Each block is integrated with coefficient q_ij.
+
+      If the vector demension does not match the true dimension of the \c
+      FiniteElementSpace , the resulting element matrix will be mathematically
+      invalid. */
+   VectorDiffusionIntegrator(MatrixCoefficient& mq) : MQ{&mq}, vdim{mq.GetVDim()} {}
```
Comment (stefanozampini) [inline fem/bilininteg.hpp:2590]: same

**Review by stefanozampini — CHANGES_REQUESTED** [review]

File: fem/bilininteg.cpp (hunk 5 above)
Comment (wcdawn) [reply fem/bilininteg.cpp:2184]: Maybe this is a confusing variable name? `vcoeff` is actually a `Vector` and *not* a `VectorCoefficient`. It is simply a variable used to store the result returned by the `VectorCoefficient`, `VQ`. I am not modifying the coefficient, just setting the size of the result `Vector` or `DenseMatrix` so that there is enough space to store the result returned by the coefficient.

File: fem/bilininteg.cpp (hunk 8 above)
Comment (wcdawn) [reply fem/bilininteg.cpp:2318]: I agree but I don't see a better way to do this. A custom `Mult_a_AAt_V` could be written but I've rewritten to at least reduce the number of operations.

File: fem/bilininteg.hpp (hunk 9 above)
Comment (wcdawn) [reply fem/bilininteg.hpp:2527]: I think it does. `vdim` is set by calling `VectorCoefficient::GetVDim()` or `MatrixCoefficient::GetVDim()` in the constructor as appropriate.

**Comment by tzanio** [comment]
Can you please merge `master` and resolve the conflicts?

**Comment by v-dobrev** [comment]
Something is wrong here: "895 commits" (new commits) does not look right.

**Comment by wcdawn** [comment]
I agree this seems pretty broken.
I tried to fix it and broke it more...

Do you have a recommendation or should I create a new PR?

**Comment by wcdawn** [comment]
I've created a new branch `nonscalar-VectorDiffusionIntegrator` (without my username) that only has one commit on it as it should. If this is acceptable, I can create a new PR for that branch.

**Comment by v-dobrev** [comment]
We can use the new branch to overwrite this branch by force pushing -- that way the PR history here will be preserved. @tzanio, do you want to do that?

**Comment by tzanio** [comment]
Ok, I force-pushed from `nonscalar-VectorDiffusionIntegrator`.

@wcdawn, can you double-check that everything looks ok?

**Comment by tzanio** [comment]
You can update your local copy of this branch with:

```
> git fetch origin wcdawn/nonscalar-VectorDiffusionIntegrator
> git checkout -B wcdawn/nonscalar-VectorDiffusionIntegrator origin/wcdawn/nonscalar-VectorDiffusionIntegrator
```

**Comment by v-dobrev** [comment]
By the way, the branch https://github.com/mfem/mfem/tree/nonscalar-VectorDiffusionIntegrator can now be deleted -- at the moment it points to the same commit as https://github.com/mfem/mfem/tree/wcdawn/nonscalar-VectorDiffusionIntegrator (this PR's branch).

**Comment by wcdawn** [comment]
Looks good to me! I've also deleted the other branch per @v-dobrev comment.
Thanks!

**Comment by tzanio** [comment]
@stefanozampini and @rw-anderson -- can you please take another look?

**Comment by tzanio** [comment]
ping: @stefanozampini, @rw-anderson

**Comment by wcdawn** [comment]
Is there any update on this? I need this for my work and having everything in `master` would make my work much easier. Thanks!

**Review by stefanozampini — APPROVED** [review]

File: fem/bilininteg.cpp (hunk 8 above)
Comment (stefanozampini) [reply fem/bilininteg.cpp]: I have accepted the PR since it does not seem you want to further optimize this and you probably do not need it.

**Review by rw-anderson — APPROVED** [review]
I made some very minor suggestions for documentation and typos. I don't have time to verify correctness, but otherwise it looks ok to me.

File: fem/bilininteg.cpp (hunk 8 above)
Comment (wcdawn) [reply fem/bilininteg.cpp]: I'm interested in further optimizing but I really don't see how without introducing a new operator somehow.

In your pseudocode, the action is
`v_o += alpha * D*D^T * v_i`
which is the product of a scalar, a matrix and its transpose, and a vector. Hence, `Mult_a_AAt_V`.

Functions like
```cpp
mfem::Vector mfem::Vector::operator*(double a);
mfem::DenseMatrix mfem::DenseMatrix::operator*(double a);
void mfem::DenseMatrix::Mult_MultTranspose(const mfem::Vector& v_i, mfem::Vector& v_o);
```
would give more options but I don't see how to implement this accumulation with the operators at hand.

Otherwise, I could use `pelmat` with `P=D*D^T` like I'm doing now and then `P.AddMult_a(w*vcoeff(k), vec_out, vec_in)`.

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: fem/bilininteg.hpp [hunk 13]
```diff
@@ -2561,17 +2569,59 @@ class VectorDiffusionIntegrator : public BilinearFormIntegrator
...
+      The element matrix is block-diagonal with \c vdim copies of the element
+      matrix integrated with the \c Coefficient.
+
+      If the vector dimension does not match the true dimension of the \c
+      FiniteElementSpace , the resulting element matrix will be mathematically
+      invalid. */
+   VectorDiffusionIntegrator(Coefficient &q, int vector_dimension)
+      : Q(&q), vdim(vector_dimension) { }
+
+   /** \brief Integrator with \c VectorCoefficient. The vector dimension of the \c FiniteElementSpace is assumed to be the same as the dimension of the \c Vector.
```
Comment (v-dobrev) [inline fem/bilininteg.hpp:2605]: Here and in some other places, the text should be wrapped at <= 80 characters.

**Review by tzanio — APPROVED** [review]
Thanks for the contribution @wcdawn !

@stefanozampini and @rw-anderson -- thank you for reviewing!
